src/Interpreters/evaluateConstantExpression.cpp

```diff
@@ -368,7 +369,30 @@ std::optional<Blocks> evaluateExpressionOverConstantCondition(const ASTPtr & nod
 
         for (const auto & conjunct : dnf)
         {
-            Block block(conjunct);
+            Block block;
+
+            for (const auto & elem : conjunct)
+            {
+                if (!block.has(elem.name))
+                {
+                    block.insert(elem);
+                }
+                else
+                {
+                    /// Conjunction of condition on column equality to distinct values can never be satisfied.
```

Comment:
@alexey-milovidov the behavior had been changed because of this hunk.

```
SELECT count(*) FROM remote('127.{1,2}', system.one, dummy) WHERE dummy = 1 AND dummy = 0 SETTINGS optimize_skip_unused_shards=1;
```

Without:

```sql
[p1.azat] 2021.07.24 06:41:06.482136 [ 3871 ] {ba1a6a15-576f-4883-9a4c-e81e60bab557} <Debug> StorageDistributed (remote): Skipping irrelevant shards - the query will be sent to the following 
shards of the cluster (shard numbers): [2]
```

With this patch:

```sql
[p1.azat] 2021.07.24 06:37:01.399750 [ 22034 ] {0aa22c9f-63e8-4630-b951-0d0b76a02c10} <Debug> StorageDistributed (remote): Skipping irrelevant shards - the query will be sent to the following shards of the cluster (shard numbers): []
```

So there are zero shards to send the query, hence nothing more is done.
Such queries does not looks like real user queries, so maybe we should not try to optimize them, or return any node if zero is returned, what do you think?

Reply:
> So there are zero shards to send the query, hence nothing more is done.

But the implementation for zero shards is wrong.
SELECT count() ... without GROUP BY returns empty set instead of value 0.